src/compiler/checker.ts

```diff
@@ -15489,13 +15489,23 @@ namespace ts {
                     }
                 }
 
-                if (isLiteralType(source) && !forEachType(target, isUnitType)) {
+                if (isLiteralType(source) && !typeCouldHaveNoTopLevelSingletonTypes(target)) {
                     generalizedSourceType = getTypeNameForErrorDisplay(getBaseTypeOfLiteralType(source));
                 }
 
                 reportError(message, generalizedSourceType, targetType);
             }
 
+            function typeCouldHaveNoTopLevelSingletonTypes(type: Type) {
+                return forEachType(type, typeCouldHaveNoTopLevelSingletonTypesWorker);
+            }
+
+            function typeCouldHaveNoTopLevelSingletonTypesWorker(type: Type): boolean {

```

Comment:
Okay, I think we're getting there. First, I would move these functions out so that they don't add more to the current closure on every call.

Second, I think we have to go deeper on constraints, and we can get rid of the "Worker" function.

```ts

            function typeCouldHaveNoTopLevelSingletonTypes(type: Type): boolean {
                if (type.flags & TypeFlags.UnionOrIntersection) {
                    return !!forEach((type as IntersectionType).types, typeCouldHaveNoTopLevelSingletonTypes);
                }

                if (type.flags & TypeFlags.Instantiable) {
                    const constraint = getConstraintOfType(type);
                    if (constraint) {
                        return typeCouldHaveNoTopLevelSingletonTypes(constraint);
                    }
                }

                return isUnitType(type);
            }
```